newrelic/resource_newrelic_alert_channel.go

```diff
@@ -86,32 +72,129 @@ func resourceNewRelicAlertChannel() *schema.Resource {
 			},
 			"configuration": {
 				Type:     schema.TypeMap,
-				Required: true,
+				Optional: true,
 				ForceNew: true,
 				//TODO: ValidateFunc: (use list of keys from map above)
-				Sensitive: true,
+				Sensitive:     true,
+				Deprecated:    "use config instead",
+				ConflictsWith: []string{"config"},
 			},
-		},
-	}
-}
 
-func buildAlertChannelStruct(d *schema.ResourceData) *newrelic.AlertChannel {
-	channel := newrelic.AlertChannel{
-		Name:          d.Get("name").(string),
-		Type:          d.Get("type").(string),
-		Configuration: d.Get("configuration").(map[string]interface{}),
+			"config": {
+				Type:          schema.TypeList,
+				Optional:      true,
+				ForceNew:      true,
+				MaxItems:      1,
+				ConflictsWith: []string{"configuration"},
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"api_key": {
+							Type:      schema.TypeString,
+							Optional:  true,
+							Sensitive: true,
+						},
+						"auth_password": {
+							Type:      schema.TypeString,
+							Optional:  true,
+							Sensitive: true,
+						},
+						"auth_type": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							Sensitive:    true,
+							ValidateFunc: validation.StringInSlice([]string{"BASIC"}, false),
+						},
+						"auth_username": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"base_url": {
+							Type:      schema.TypeString,
+							Optional:  true,
+							Sensitive: true,
+						},
+						"channel": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"headers": {
+							Type:      schema.TypeMap,
+							Elem:      &schema.Schema{Type: schema.TypeString},
+							Optional:  true,
+							Sensitive: true,
+						},
+						"key": {
+							Type:      schema.TypeString,
+							Optional:  true,
+							Sensitive: true,
+						},
+						"include_json_attachment": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"payload": {
+							Type:      schema.TypeMap,
+							Elem:      &schema.Schema{Type: schema.TypeString},
+							Sensitive: true,
+							Optional:  true,
+						},
+						"payload_type": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"recipients": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"region": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: validation.StringInSlice([]string{"US", "EU"}, false),
```

Comment:
Nice.   Might be a good place to add `newrelic.Region.Names` or something that returns a `[]string` of the valid regions

Reply:
This argument is specific to the OpsGenie alert channel, though.  There isn't a supported "staging" environment, for instance.